### PR TITLE
Replace assert with KOKKOS_ASSERT in kernel code

### LINKED ISSUE #949: Replace use of the macro `assert` in kernels
We need to replace our use of the `assert` macro from `<cassert>` in kernels by `KOKKOS_ASSERT` or something equivalent.
@mrlag31 reported issues with Clang+CUDA on his local machine in Debug mode.
Our CI build happens to be a Debug build but these are using different versions of Clang and CUDA.

### PR DESCRIPTION
**Pull request by masterleinad**
Fixes #949.

### CODE REVIEW

**Review by aprokop — COMMENTED** [review]
Might as well also remove the `<cassert>` header inclusion in files that don't use `assert`.

**Comment by aprokop** [comment]
Would be nice if kokkos/kokkos#5403 was resolved first to not bring `Kokkos_Core.hpp`.

**Comment by dalg24** [comment]
Why is it acceptable to pull in the full `<Kokkos_Core.hpp>` header?
What are the pros and cons as compared to rolling our own?

**Comment by masterleinad** [comment]
> Why is it acceptable to pull in the full `<Kokkos_Core.hpp>` header? What are the pros and cons as compared to rolling our own?

My two cents:
- Effectively, any code that uses `ArborX` needs `Kokkos_Core.hpp` anyway since that's what you need for a `parallel_for` or for using `View` anyway as of now. It's realistic to make the needed `Kokkos` header public pretty soon and transitioning is very easy since it only means to change an include. 
- It seems that `KOKKOS_ASSERT` is doing what we want and so coming up with our own solution/macro would probably be pretty much a copy of what `Kokkos` provides already. If we roll or own, potentially moving to `KOKKOS_ASSERT` when we only need the `Kokkos_Error.hpp` (or whatever it will be called) header requires more changes than using it from the beginning.

**Review by aprokop — COMMENTED** [review]
Should we create `ArborX_KokkosExtAssert.hpp` to hide the complexity?

File: src/details/ArborX_DetailsCartesianGrid.hpp [hunk 1]
```diff
@@ -17,10 +17,13 @@
 #include <ArborX_HyperBox.hpp>
 
 #include <Kokkos_Macros.hpp>
+#if KOKKOS_VERSION >= 40199
```
Comment (dalg24) [inline src/details/ArborX_DetailsCartesianGrid.hpp:20]: kokkos/kokkos#6445 was merged relatively late after we branched and release 4.1
```suggestion
#if KOKKOS_VERSION >= 40200
```
It is safer to guard against 4.2

Comment (aprokop) [reply src/details/ArborX_DetailsCartesianGrid.hpp:20]: Agree with that.

File: src/details/ArborX_DetailsCartesianGrid.hpp [hunk 2]
```diff
@@ -17,10 +17,13 @@
 #include <ArborX_HyperBox.hpp>
 
 #include <Kokkos_Macros.hpp>
+#if KOKKOS_VERSION >= 40200
+#include <Kokkos_Assert.hpp> // KOKKOS_ASSERT
+#else
+#include <Kokkos_Core.hpp> // KOKKOS_ASSERT
```
Comment (dalg24) [inline src/details/ArborX_DetailsCartesianGrid.hpp:23]: Did you consider including `impl/Kokkos_Error.hpp` instead?  I suspect that would work.

Comment (masterleinad) [reply src/details/ArborX_DetailsCartesianGrid.hpp:23]: Cons:
- `impl/Kokkos_Error.hpp` is not a public header. We shouldn't be a bad example.
- up to `Kokkos` version 4.2.0 (at the very least), there is not much of a chance for an `ArborX` executable to be able to include less than `Kokkos_Core.hpp` anyway.

Pros:
- we only include non-public headers up to a version that we know the file exists.
- we don't rely on `Kokkos_Core.hpp` including headers we should provide ourselves, see https://github.com/arborx/ArborX/pull/956, before version 4.2.0.

Comment (dalg24) [reply src/details/ArborX_DetailsCartesianGrid.hpp:23]: > Cons:
> 
> * `impl/Kokkos_Error.hpp` is not a public header. We shouldn't be a bad example.

In our implementation details only.  I don't think that is an issue.

> * up to `Kokkos` version 4.2.0 (at the very least), there is not much of a chance for an `ArborX` executable to be able to include less than `Kokkos_Core.hpp` anyway.

I don't understand the point you want to make.

**Comment by masterleinad** [comment]
Retest this please.

**Review by dalg24 — APPROVED** [review]

**Review by aprokop — APPROVED** [review]

**Comment by aprokop** [comment]
HIP did not start, one CUDA build ran out of disk space. Otherwise, good.

**Pull request merged**